snuba/query/query.py

```diff
@@ -57,6 +104,57 @@ def __init__(self, body: MutableMapping[str, Any], data_source: RelationalSource
         self.__data_source = data_source
         self.__prewhere_conditions: Sequence[Condition] = []
 
+        # New data model
+        # TODO: Provide a better typing for this.
+        self.__selected_columns = selected_columns or []
+        self.__array_join = array_join
+        self.__condition = condition
+        self.__groupby = groupby or []
+        self.__having = having
+        self.__order_by = order_by or []
+
+    def get_all_expressions(self) -> Iterable[Expression]:
+        """
+        Returns an expression container that iterates over all the expressions
+        in the query no matter which level of nesting they are at.
+        The ExpressionContainer can be used to traverse the expressions in the
+        tree.
+        """
+        return chain(
+            chain.from_iterable(self.__selected_columns),
+            self.__array_join or [],
+            self.__condition or [],
+            chain.from_iterable(self.__groupby),
+            self.__having or [],
+            chain.from_iterable(map(lambda orderby: orderby.node, self.__order_by)),
+        )
+
+    def transform_expressions(
+        self,
+        func: Callable[[Expression], Expression],
+    ) -> None:
+        """
+        Transforms in place the current query object by applying a transformation
+        function to all expressions contained in this query
+        """
```

Comment:
Should this return a new `Query` to mirror the semantics of `transform`?

Reply:
I wish , unfortunately we cannot do that just yet:
There are several places during the query execution that rely on Request being a sealed dataclass with a query argument, thus We cannot swap the Query as of now. 
This is what drove the QueryProcessor to make changes to the query object in place.

We need first to remove all dependencies on a mutable Query, then we can make Query immutable itself.

Reply:
That makes sense. It'd probably a good thing to note in a comment that this was a design decision made as a side effect of another design decision (so we don't have to remember why it was written this way if/when we can change it later.)